Добавлены реализации parcel.go и parcel_test.go

Alex-bot-278 opened the pull request.

Alex-bot-278 committed:
Добавлены реализации parcel.go и parcel_test.go

dirtymew commented on parcel.go:
```diff
@@ -14,47 +14,104 @@ func NewParcelStore(db *sql.DB) ParcelStore {
 	p := Parcel{}
+	err := row.Scan(&p.Number, &p.Client, &p.Status, &p.Address, &p.CreatedAt)
+	if err != nil {
+		return p, err
```
return Parcel{}, err

dirtymew commented on parcel_test.go:
```diff
@@ -28,60 +24,92 @@ func getTestParcel() Parcel {
-	// проверьте, что значения всех полей в полученном объекте совпадают со значениями полей в переменной parcel
+	storedParcel, err := store.Get(id)
+	require.NoError(t, err)
+	require.Equal(t, parcel.Client, storedParcel.Client)
```
assert

dirtymew replied:
поле Number тоже нужно проверять

dirtymew commented on parcel_test.go:
```diff
@@ -28,60 +24,92 @@ func getTestParcel() Parcel {
+	require.NoError(t, err)
+
+	_, err = store.Get(id)
+	require.ErrorIs(t, err, sql.ErrNoRows)
```
require.Error
assert.ErrorIs

dirtymew commented on parcel_test.go:
```diff
@@ -28,60 +24,92 @@ func getTestParcel() Parcel {
-	// получите добавленную посылку и убедитесь, что адрес обновился
+	storedParcel, err := store.Get(id)
+	require.NoError(t, err)
+	require.Equal(t, newAddress, storedParcel.Address)
```
assert

dirtymew commented on parcel_test.go:
```diff
@@ -28,60 +24,92 @@ func getTestParcel() Parcel {
-	// получите добавленную посылку и убедитесь, что статус обновился
+	storedParcel, err := store.Get(id)
+	require.NoError(t, err)
+	require.Equal(t, ParcelStatusSent, storedParcel.Status)
```
assert

dirtymew commented on parcel_test.go:
```diff
@@ -90,32 +118,33 @@ func TestGetByClient(t *testing.T) {
-	// убедитесь, что количество полученных посылок совпадает с количеством добавленных
+	storedParcels, err := store.GetByClient(client)
+	require.NoError(t, err)
+	require.Len(t, storedParcels, len(parcels))
```
assert

dirtymew commented on parcel_test.go:
```diff
@@ -90,32 +118,33 @@ func TestGetByClient(t *testing.T) {
-		// убедитесь, что все посылки из storedParcels есть в parcelMap
-		// убедитесь, что значения полей полученных посылок заполнены верно
+		expectedParcel, ok := parcelMap[parcel.Number]
+		require.True(t, ok)
```
assert

dirtymew commented on parcel_test.go:
```diff
@@ -90,32 +118,33 @@ func TestGetByClient(t *testing.T) {
-		// убедитесь, что значения полей полученных посылок заполнены верно
+		expectedParcel, ok := parcelMap[parcel.Number]
+		require.True(t, ok)
+		require.Equal(t, expectedParcel.Client, parcel.Client)
```
require.Equal(t, expectedParcel, parcel)

dirtymew reviewed (changes requested):
Привет Александр

сервисе не работает - нужно доделать работу см main.go

сразу дал комментарии что поправить по коду и тестам

Alex-bot-278 committed:
fix(parcel): исправлены методы Get, SetAddress и Delete согласно ревью

dirtymew commented on main.go:
```diff
@@ -40,7 +40,7 @@ func (s ParcelService) Register(client int, address string) (Parcel, error) {
 	id, err := s.store.Add(parcel)
 	if err != nil {
-		return parcel, err
+		return Parcel{}, fmt.Errorf("ошибка регистрации посылки: %w", err)
```
ошибки на английском принято писать

dirtymew commented on parcel.go:
```diff
@@ -13,48 +15,161 @@ func NewParcelStore(db *sql.DB) ParcelStore {
+	err := row.Scan(&p.Number, &p.Client, &p.Status, &p.Address, &p.CreatedAt)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return Parcel{}, fmt.Errorf("посылка с номером %d не найдена", number)
```
так не стоит делать - ты теряешь изначальную ошибку - тогда лучше обернуть если хочется добавить контекста

dirtymew commented on parcel.go:
```diff
@@ -13,48 +15,161 @@ func NewParcelStore(db *sql.DB) ParcelStore {
 func (s ParcelStore) SetAddress(number int, address string) error {
-	// реализуйте обновление адреса в таблице parcel
-	// менять адрес можно только если значение статуса registered
+	tx, err := s.db.Begin()
```
супер что погрузился в транзакции - но тут это перебор

нужно написать один запрос UPDATE с правильным условием

если сможешь обосновать почему тут стоит делать это в транзакции, то приму и так - если нет тогда нужно переделать

dirtymew commented on parcel.go:
```diff
@@ -13,48 +15,161 @@ func NewParcelStore(db *sql.DB) ParcelStore {
 func (s ParcelStore) Delete(number int) error {
-	// реализуйте удаление строки из таблицы parcel
-	// удалять строку можно только если значение статуса registered
+	tx, err := s.db.Begin()
```
тоже замечание что и для SetAddress

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
 
-// TestAddGetDelete проверяет добавление, получение и удаление посылки
+func setupDB(t *testing.T) *sql.DB {
+	db, err := sql.Open("mysql", "user:password@/dbname")
```
эти тесты не работают - у меня нет mysql 

поэтому нужно сделать на sqlite - либо  tracker.db либо отдельный файл для тестов или в памяти

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	require.NoError(t, err)
+
+	// Очищаем таблицу перед тестами
+	_, err = db.Exec("DELETE FROM parcel")
```
DELETE * или TRANCATE

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// проверьте, что посылку больше нельзя получить из БД
+	// Add
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	// Add
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
+	assert.NotZero(t, id)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+
+	// Get
+	storedParcel, err := store.Get(id)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+
+	// Verify delete
+	_, err = store.Get(id)
+	assert.Error(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	// Verify delete
+	_, err = store.Get(id)
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, sql.ErrNoRows)
```
см 44 строку parcel.go - этот тест упадет

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// добавьте новую посылку в БД, убедитесь в отсутствии ошибки и наличии идентификатора
+	// Add
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	// Set address
 	newAddress := "new test address"
+	err = store.SetAddress(id, newAddress)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// получите добавленную посылку и убедитесь, что адрес обновился
+	// Check
+	updatedParcel, err := store.Get(id)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// получите добавленную посылку и убедитесь, что статус обновился
+	// Check
+	updatedParcel, err := store.Get(id)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// добавьте новую посылку в БД, убедитесь в отсутствии ошибки и наличии идентификатора
+	// Add
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	// Set status
+	newStatus := ParcelStatusSent
+	err = store.SetStatus(id, newStatus)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
 	for i := 0; i < len(parcels); i++ {
-		id, err := // добавьте новую посылку в БД, убедитесь в отсутствии ошибки и наличии идентификатора
+		id, err := store.Add(parcels[i])
+		assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-		id, err := // добавьте новую посылку в БД, убедитесь в отсутствии ошибки и наличии идентификатора
+		id, err := store.Add(parcels[i])
+		assert.NoError(t, err)
+		assert.NotZero(t, id)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
-	// убедитесь, что количество полученных посылок совпадает с количеством добавленных
+	// Get by client
+	storedParcels, err := store.GetByClient(client)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+
+	// Add and set status to sent
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
+	err = store.SetStatus(id, ParcelStatusSent)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+
+	// Try to set address
+	err = store.SetAddress(id, "new address")
+	assert.Error(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	// Try to set address
+	err = store.SetAddress(id, "new address")
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "нельзя изменить адрес")
```
для ошибко есть функции ErrorEqual или ErrorIs

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+
+	// Add and set status to sent
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
```
require

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,169 @@ func getTestParcel() Parcel {
+	id, err := store.Add(parcel)
+	assert.NoError(t, err)
+	err = store.SetStatus(id, ParcelStatusSent)
+	assert.NoError(t, err)
```
require

dirtymew commented on main.go:
```diff
@@ -137,7 +167,6 @@ func main() {
 	err = service.Delete(p.Number)
 	if err != nil {
 		fmt.Println(err)
-		return
```
ок )

dirtymew reviewed (changes requested):
Александр

по работе есть замечания и вопросы

Alex-bot-278 committed:
Fix all review comments:
- Translate error messages to English
- Wrap errors properly with %w
- Simplify SetAddress and Delete without transactions
- Fix tests to use SQLite in memory
- Proper require/assert usage in tests
- Better error checking with ErrorIs

Alex-bot-278 commented:
Перевод ошибок на английский;   Правильное оборачивание ошибок с %w;
Упрощение SetAddress/Delete (без транзакций);
 Исправление тестов (SQLite in memory);
Правильное использование require/assert;
 Улучшенная проверка ошибок в тестах.

dirtymew commented on parcel_test.go:
```diff
@@ -28,94 +25,156 @@ func getTestParcel() Parcel {
+	require.NoError(t, err)
+
+	err = store.Delete(id)
+	assert.Error(t, err)
```
require

dirtymew reviewed (approved):
Александр

работа принята
